flink-kubernetes/src/main/java/org/apache/flink/kubernetes/KubernetesResourceManagerDriver.java

```diff
@@ -0,0 +1,325 @@
+		final List<KubernetesWorkerNode> recoveredWorkers = new ArrayList<>();
+
+		for (KubernetesPod pod : podList) {
+			final KubernetesWorkerNode worker = new KubernetesWorkerNode(new ResourceID(pod.getName()));
+			recoveredWorkers.add(worker);
+			final long attempt = worker.getAttempt();
+			if (attempt > currentMaxAttemptId) {
+				currentMaxAttemptId = attempt;
+			}
+		}
+
+		log.info("Recovered {} pods from previous attempts, current attempt id is {}.",
+				recoveredWorkers.size(),
+				++currentMaxAttemptId);
+
+		// Should not invoke resource event handler on the main thread executor.
+		// We are in the initializing thread. The main thread executor is not yet ready.
+		getResourceEventHandler().onPreviousAttemptWorkersRecovered(recoveredWorkers);
+	}
+
+	private KubernetesTaskManagerParameters createKubernetesTaskManagerParameters(TaskExecutorProcessSpec taskExecutorProcessSpec) {
+		final String podName = String.format(
+				TASK_MANAGER_POD_FORMAT,
+				clusterId,
+				currentMaxAttemptId,
+				++currentMaxPodId);
+
+		final ContaineredTaskManagerParameters taskManagerParameters =
+				ContaineredTaskManagerParameters.create(flinkConfig, taskExecutorProcessSpec);
+
+		final Configuration taskManagerConfig = new Configuration(flinkConfig);
+		taskManagerConfig.set(TaskManagerOptions.TASK_MANAGER_RESOURCE_ID, podName);
+
+		final String dynamicProperties =
+				BootstrapTools.getDynamicPropertiesAsString(flinkClientConfig, taskManagerConfig);
+
+		return new KubernetesTaskManagerParameters(
+				flinkConfig,
+				podName,
+				dynamicProperties,
+				taskManagerParameters,
+				ExternalResourceUtils.getExternalResources(flinkConfig, KubernetesConfigOptions.EXTERNAL_RESOURCE_KUBERNETES_CONFIG_KEY_SUFFIX));
+	}
+
+	private void tryResetPodCreationCoolDown() {
+		if (podCreationCoolDown.isDone()) {
+			log.info("Pod creation failed. Will not retry creating pods in {}.", podCreationRetryInterval);
+			podCreationCoolDown = new CompletableFuture<>();
+			getMainThreadExecutor().schedule(
+					() -> podCreationCoolDown.complete(null),
+					podCreationRetryInterval.getSize(),
+					podCreationRetryInterval.getUnit());
+		}
+	}
+
+	private void terminatedPodsInMainThread(List<KubernetesPod> pods) {
+		getMainThreadExecutor().execute(() -> {
+			for (KubernetesPod pod : pods) {
+				if (pod.isTerminated()) {
+					final String podName = pod.getName();
+					log.info("TaskManager pod {} is terminated.", podName);
+
+					// this is a safe net, in case onModified/onDeleted/onError is received before onAdded
+					final CompletableFuture<KubernetesWorkerNode> requestResourceFuture = requestResourceFutures.remove(podName);
+					if (requestResourceFuture != null) {
+						log.warn("Pod {} is terminated before receiving the ADDED event.", podName);
+						requestResourceFuture.completeExceptionally(new FlinkException("Pod is terminated."));
+					}
+
+					getResourceEventHandler().onWorkerTerminated(new ResourceID(podName));
```

Comment:
Do we need to call `closeTaskManagerConnection` in `ActiveResourceManager#onWorkerTerminated`?

Reply:
I believe there is another ticket (FLINK-17273) tracking this issue.
Since this is also a behavior change, and is not necessarily related to this refracting effort, I would be in favor of making it in a separate PR.